src/dos/cdrom_image.cpp

```diff
@@ -249,51 +249,102 @@ bool CDROM_Interface_Image::AudioFile::read(uint8_t *buffer,
+	const uint8_t channels = getChannels();
+	const uint8_t bytes_per_frame = channels * REDBOOK_BPS;
+	const uint32_t requested_frames = ceil_divide(requested_bytes, BYTES_PER_REDBOOK_PCM_FRAME);
 
+	uint32_t decoded_bytes = 0;
 	uint32_t decoded_frames = 0;
-	const uint32_t requested_frames = requested_bytes / BYTES_PER_REDBOOK_PCM_FRAME;
 	while (decoded_frames < requested_frames) {
-		if (sample->flags & (SOUND_SAMPLEFLAG_ERROR | SOUND_SAMPLEFLAG_EOF))
+		const uint32_t decoded = Sound_Decode_Direct(sample,
+		                                             buffer + decoded_bytes,
+		                                             requested_frames - decoded_frames);
+		if (sample->flags & (SOUND_SAMPLEFLAG_ERROR | SOUND_SAMPLEFLAG_EOF) || !decoded)
 			break;
-		decoded_frames += Sound_Decode_Direct(sample,
-		                                      buffer + decoded_frames * BYTES_PER_REDBOOK_PCM_FRAME,
-		                                      requested_frames - decoded_frames);
+		decoded_frames += decoded;
+		decoded_bytes = decoded_frames * bytes_per_frame;
 	}
+	// Zero out any remainining frames that we didn't fill
 	if (decoded_frames < requested_frames)
-		memset(buffer + decoded_frames * BYTES_PER_REDBOOK_PCM_FRAME,
-		       0,
-		       (requested_frames - decoded_frames) * BYTES_PER_REDBOOK_PCM_FRAME);
+		memset(buffer + decoded_bytes, 0, requested_bytes - decoded_bytes);
+
+	// If the track is mono, convert to stereo
+	if (channels == 1 && decoded_frames) {
+#ifdef DEBUG
+		using namespace std::chrono;
+		using clock = std::chrono::steady_clock;
+		clock::time_point begin = clock::now(); // start the timer
+#endif
+		/**
+		 *  Convert to stereo in-place:
+		 *  The current buffer is half full of mono samples:
+		 *  0. [mmmmmmmm00000000]
+		 * 
+		 *  We walk backward from the last mono sample to the first,
+		 *  duplicating them to left and right samples at the rear
+		 *  of the buffer:
+		 *  1. [mmmmmmm(m)000000(LR)]
+		 *              ^________\/
+		 * 
+		 *  2. [mmmmmm(m)m0000(LR)LR]
+		 *             ^_______\/
+		 * 
+		 *  Eventually the left and right samples overwrite the prior
+		 *  mono samples until the buffer is full of LR samples (where
+		 *  the first mono samples is simply the 'left' value):
+		 * 
+		 *  (N - 1). [(m)(R)LRLRLRLRLRLRLR]
+		 *             ^_/
+		 */
+		int16_t* pcm_buffer = reinterpret_cast<int16_t*>(buffer);
+		const uint32_t mono_samples = decoded_frames;
+		for (uint32_t i = mono_samples - 1; i > 0; --i) {
+			pcm_buffer[i * REDBOOK_CHANNELS + 1] = pcm_buffer[i]; // right
+			pcm_buffer[i * REDBOOK_CHANNELS + 0] = pcm_buffer[i]; // left
+		}
+
+#ifdef DEBUG
+		clock::time_point end = clock::now(); // stop the timer
+		const int32_t elapsed_us = static_cast<int32_t>
+		    (duration_cast<microseconds>(end - begin).count());
+		LOG_MSG("CDROM: converted %u mono to %u samples in %d us",
+	             mono_samples, mono_samples * REDBOOK_CHANNELS, elapsed_us);
```

Comment:
Interesting; I suppose all of the fixed-size types map to an equivalent PR* print string.  Yes; I agree.. and the fact that neither gcc or Clang warn about this "future size change risk", even at their maximum warning sensitivity, says that that using `%d / %u` and `%ld / %lu` are essentially as reliable as the language specification.

Reply:
Compilers can't warn about it, because PRI* macros are expanded by preprocessor - they literally don't see the difference - at least as long as all compilers use the same size for int (which they do, AFAIK - but language specification does not require int to be 32-bit). So `%d` for `int32_t` will work. But `%ld` for `int64_t` won't - because at least on Windows `long` is the same size as `int` :/ (that's why I replaced `Bitu` with `unsigned` in one of recent rounds of fixes for formatting warnings).

In this specific case I think maybe `int` could've been used instead of `int32_t`, but it's a non-issue to me - we can leave it as it is.